pkg/app/app_list_test.go

```diff
@@ -0,0 +1,249 @@
+package app
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"path/filepath"
+	"sync"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/assert"
+	"github.com/variantdev/vals"
+
+	"github.com/helmfile/helmfile/pkg/helmexec"
+	"github.com/helmfile/helmfile/pkg/testhelper"
+	"github.com/helmfile/helmfile/pkg/testutil"
+)
+
+func TestListWithEnvironment(t *testing.T) {
+	type testcase struct {
+		environment string
+		ns          string
+		error       string
+		selectors   []string
+		expected    string
+	}
+
+	check := func(t *testing.T, tc testcase) {
+		t.Helper()
+
+		bs := &bytes.Buffer{}
+
+		func() {
+			t.Helper()
+
+			logReader, logWriter := io.Pipe()
+
+			logFlushed := &sync.WaitGroup{}
+			// Ensure all the log is consumed into `bs` by calling `logWriter.Close()` followed by `logFlushed.Wait()`
+			logFlushed.Add(1)
+			go func() {
+				scanner := bufio.NewScanner(logReader)
+				for scanner.Scan() {
+					bs.Write(scanner.Bytes())
+					bs.WriteString("\n")
+				}
+				logFlushed.Done()
+			}()
+
+			defer func() {
+				// This is here to avoid data-trace on bytes buffer `bs` to capture logs
+				if err := logWriter.Close(); err != nil {
+					panic(err)
+				}
+				logFlushed.Wait()
+			}()
+
+			logger := helmexec.NewLogger(logWriter, "debug")
+
+			valsRuntime, err := vals.New(vals.Options{CacheSize: 32})
+			if err != nil {
+				t.Errorf("unexpected error creating vals runtime: %v", err)
+			}
+
+			files := map[string]string{
+				"/path/to/helmfile.d/helmfile_1.yaml": `
+environments:
+  development: {}
+  shared: {}
+
+releases:
+- name: logging
+  chart: incubator/raw
+  namespace: kube-system
+
+- name: kubernetes-external-secrets
+  chart: incubator/raw
+  namespace: kube-system
+  needs:
+  - kube-system/logging
+
+- name: external-secrets
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: test
+  needs:
+  - kube-system/kubernetes-external-secrets
+
+- name: my-release
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: test
+  needs:
+  - default/external-secrets
+
+
+# Disabled releases are treated as missing
+- name: disabled
+  chart: incubator/raw
+  namespace: kube-system
+  installed: false
+
+- name: test2
+  chart: incubator/raw
+  needs:
+  - kube-system/disabled
+
+- name: test3
+  chart: incubator/raw
+  needs:
+  - test2
+`,
+				"/path/to/helmfile.d/helmfile_2.yaml": `
+environments:
+  test: {}
+  shared: {}
+
+repositories:
+- name: bitnami
+  url: https://charts.bitnami.com/bitnami
+
+releases:
+- name: cache
+  namespace: my-app
+  chart: bitnami/redis
+  version: 17.0.7
+  labels:
+    app: test
+
+- name: database
+  namespace: my-app
+  chart: bitnami/postgres
+  version: 11.6.22
+`,
+				"/path/to/helmfile.d/helmfile_3.yaml": `
+releases:
+- name: global
```

Comment:
Another thing worth calling out is that this `global` release can never be included when an environment (other than `default`) is specified. This is consistent with the earlier behaviour, but I'm unclear whether it is intended, or whether such releases should be included for _every_ environment?

Reply:
@dackroyd Good catch! That's how it's intended to work today.

Suppose we only have two envs in this specific example, it's a matter of adding:

```
environments:
  test: {}
  shared: {}
```

to `helmfile_3.yaml` here to achieve enable to for every env.

There's no easy way to make it dynamically adopt any environments defined upstream. I'm open to implementing such a feature as long as it's backward compatible.

A potential enhancement would be to make it possible to declare a helmfile.yaml to be enabled for every possible env, like:

```
environments:
  "*": {}
```

or maybe

```
environments:
  $any: {}
```

This way, we have all the syntaxes for all the use-cases:

(1) If you want to enable it only for certain envs, do write:

```
environments:
  test: {}
  shared: {}
```

(2) If you want to enable it or every env, do write:

```
environments:
  $any: {}
```

(3) If you want to enable it only for the `default` env, do omit the entire `environments` section, or:

```
environments:
  default: {}
```

Does the whole idea make sense to you?

If so, what's your preference between `"*": {}` and `$any: {}`?
I personally prefer the latter because I think it's more clear and more open to extension(a potential extension to it would be `$inherit: {}` that inherits all the environments defined in the caller helmfile.yam...), but I'm still interested in your idea.